src/main/scala/firrtl/RenameMap.scala

```diff
@@ -401,7 +409,19 @@ final class RenameMap private (val underlying: mutable.HashMap[CompleteTarget, S
     if(getCache.contains(key)) {
       getCache(key)
     } else {
-      val getter = recursiveGet(errors)(_)
+
+      // Used to determine if renaming has ended early. This is to prevent the case
+      //   where a fully recursive renaming occurs, e.g.
+      //   Given Renames:
+      //     - (~Top|Top/a:A -> ~Top|A)
+      //     - (~Top|A -> ~Top|B)
+      //   Renaming (~Top|Top/a:A) should return (~Top|A), not (~Top|B)
+      // This is due to our splitting recursive renaming into three parts (top, mid, bot)
+      def continueRenaming[T <: CompleteTarget](original: Seq[CompleteTarget], ts: Seq[CompleteTarget]): Boolean =
+        ts match {
+          case Seq(_: T) if original != ts => false
+          case _ => true
+        }
```

Comment:
self-renames should stop renaming as well right? this looks like it would return `true` on self-rename.

Reply:
I added some clarifying comments, but this code is used to stop the partial renaming from occurring. If a partial rename is a self-rename, then it shouldn't necessarily stop because sub-parts of it could be renamed. This is really only necessary for examples like the one in the comment, where you get a circular rename if you rename InstanceTarget -> ModuleTarget, then ModuleTarget -> ModuleTarget.